Sources/NIO/SocketChannel.swift

```diff
@@ -520,6 +548,47 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
         }
         return readResult
     }
+    
+    private struct ControlMessageReceiver {
+        var ecnValue: NIOExplicitCongestionNotificationState = .transportNotCapable // Default
+        
+        mutating func receiveMessage(level: Int32, type: Int32, data: UnsafeBufferPointer<UInt8>?) {
+            #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+            let ipv4TosType = IP_RECVTOS
+            #else
+            let ipv4TosType = IP_TOS    // Linux
+            #endif
+            if level == IPPROTO_IP && type == ipv4TosType {
+                if let data = data {
+                    assert(data.count == 1)
```

Comment:
Can these `assert`s have companion `precondition`s that enforce that `data.count` is _at least_ the size we want? Having the exact assert is good for debug time to spot assumptions in our code, but the `precondition` should be present so that instead of invalid reads we crash.